src/Database/LSMTree/Internal/RawPage.hs

```diff
@@ -153,26 +164,90 @@ rawPageLookup !page !key
+--
+--   * maybe (maximum (rawPageKeys page) < key) (rawPageFindKey page key)
+rawPageFindKey
+    :: RawPage
+    -> SerialisedKey
+    -> Maybe Word16  -- ^ entry number of first entry greater or equal to the key
+rawPageFindKey !page !key
+  | dirNumKeys == 1 = lookup1
+  | otherwise       = case bisectPageToKey page key of
+    KeyNotInPage   -> Nothing
+    KeyFoundAt _ i -> Just $ fromIntegral i
+
+  where
+    !dirNumKeys = rawPageNumKeys page
+
+    lookup1
+      | key <= rawPageKeyAt page 0 = Just 0
+      | otherwise                  = Nothing
+
+-- | This data-type facilitates the code reuse of 'bisectPageToKey' between the
+-- and 'rawPageLookup' and 'rawPageFindKey'.
+data BinarySearchResult
+  = KeyNotInPage
+  | KeyFoundAt !Bool {-# UNPACK #-} !Int
+
+-- | Binary search procedure shared between 'rawPageLookup' and 'rawPageFindKey'.
+{-# INLINE bisectPageToKey #-}
+bisectPageToKey
+  :: RawPage
+  -> SerialisedKey
+  -> BinarySearchResult
+bisectPageToKey !page !key = go 0 quantity
+  where
+    quantity = fromIntegral $ rawPageNumKeys page
+    -- Binary search for within the range [i, j)
+    -- for the index k corresponding to key.
+    --
+    -- If k > key, search [ k + 1, j )
+    -- If k < key, search [     i, j )
+    go :: Int -> Int -> BinarySearchResult
+    go !i !j
+      | j - i < threshold = linear i j
+      | otherwise =
+        let !k = i + div2 (j - i)
+        in  case key `compare` rawPageKeyAt page k of
+          GT -> go (k + 1) j
+          EQ -> KeyFoundAt True k
+          LT -> go i k
+
     -- when to switch to linear scan
     -- this a tuning knob
     -- can be set to zero.
     threshold = 3
 
-    bisect :: Int -> Int -> RawPageLookup (Entry SerialisedValue BlobSpan)
-    bisect !i !j
-        | j - i < threshold = linear i j
-        | otherwise = case compare key (rawPageKeyAt page k) of
-            EQ -> LookupEntry (rawPageEntryAt page k)
-            GT -> bisect (k + 1) j
-            LT -> bisect i k
-      where
-        k = i + div2 (j - i)
-
-    linear :: Int -> Int -> RawPageLookup (Entry SerialisedValue BlobSpan)
+    -- Search for an index k satisfying:
+    --   * k   in [i, j)
+    --   * k   >= key
+    --   * k-1 <  key
+    --
+    -- Note that when ending the linear scan, we must be careful to test the boundary!
+    -- If the end of the linear scan is the last key, then we are done and the key is not found.
+    -- If there is another key after the linear scan area,
+    -- then we must perform a final comparison to ensure that the key of boundary index is not
+    -- the smallest key which is greater than the search key.
+    {-# INLINE linear #-}
+    linear :: Int -> Int -> BinarySearchResult
     linear !i !j
-        | i >= j                     = LookupEntryNotPresent
-        | key == rawPageKeyAt page i = LookupEntry (rawPageEntryAt page i)
-        | otherwise                  = linear (i + 1) j
+      | i > j || (i == j && i == quantity) = KeyNotInPage
```

Comment:
`i > j` should never happen, so no need to check it. We could add `i <= j` as an invariant in the comment above and also an assertion.

Also, if `i == j` but `i /= quantity`, we know immediately that `KeyFoundAt False j` should be returned (the range `[i, j)` is empty), no need to call `rawPageKeyAt` again.

So overall this might look like `| i == j = if i == quantity then KeyNotInPage else KeyFoundAt False j`.

Reply:
You are correct, that is equivalent and simpler logic. I will leave the `i >= j` for the sake of ensuring totality of the stopping condition though. You never know when a cosmic ray might flip a bit and we need the loop to stop 😉.